src/Polly.Shared/Monkey/MonkeyEngine.cs

```diff
@@ -0,0 +1,51 @@
+﻿using System;
+using System.Threading;
+
+namespace Polly.Monkey
+{
+    internal static partial class MonkeyEngine
+    {
+        private static Random rand = new Random();
+
+        internal static TResult Implementation<TResult>(
+            Func<Context, CancellationToken, TResult> action,
+            Context context,
+            CancellationToken cancellationToken,
+            Action<Context> fault,
+            Func<Context, Decimal> injectionRate,
+            Func<Context, bool> enabled)
+        {
+            if (enabled(context) && GetRandomNumber() < injectionRate(context))
+            {
+                fault(context);
+            }
+
+            return action(context, cancellationToken);
+        }
+
+        internal static TResult Implementation<TResult>(
+            Func<Context, CancellationToken, TResult> action,
+            Context context,
+            CancellationToken cancellationToken,
+            Func<Context, Exception> fault,
```

Comment:
Probably want something like `Func<Context, CancellationToken, DelegateResult<TResult>>`, to enable this to return `TResult` values as faults (see notes in syntax classes), 

([`DelegateResult<TResult>`](https://github.com/App-vNext/Polly/blob/7dff24ffab80770e52c1b6d54a12d8fb9a7816bb/src/Polly.Shared/DelegateResult.cs) is a "poor-man's" discriminated union between a `TResult` and an `Exception`.)

Reply:
Implemented below. I have a question: 

One way of injecting the fault is to inject any generic type for example `TResult` i.e. the same return type as expected from original action being executed as you mentioned here.

Other way is to expect it to fail with an exception where the expected outcome of the original action is still `TResult`. And these seem as two disjoint scenarios. So rather than using overload I introduced here wouldn't it be better to have one like this:

```c#
internal static TResult Implementation<TResult>(
            Func<Context, CancellationToken, TResult> action,
            Context context,
            CancellationToken cancellationToken,
            Func<Context, TResult> fault,
            Func<Context, Double> injectionRate,
            Func<Context, bool> enabled)
        {
            if (enabled(context) && GetRandomNumber() < injectionRate(context))
            {
                return fault(context);
            }
             return action(context, cancellationToken);
        }
```

Reply:
@mebjas Re:

>Other way is to expect it to fail with an exception where the expected outcome of the original action is still TResult. And these seem as two disjoint scenarios. 
>So rather than using overload I introduced here wouldn't it be better to have ...

Yes, that is fine.

Reply:
Implemented